### ISSUE TITLE
event.requestContext.path.slice is not a function

### ISSUE THREAD

**Issue by adamwilbert** [dougmoscrop/serverless-http#89]
Attempting to use the newer `async: true` for event based function invocations (auto reply 200, async processing afterward in lambda) with serverless 1.41.1 and getting an error when invoking some async functions.

```
{"errorMessage":"event.requestContext.path.slice is not a function","errorType":"TypeError","stackTrace":["module.exports (/var/task/node_modules/serverless-http/lib/provider/aws/create-request.js:35:45)","/var/task/node_modules/serverless-http/lib/provider/aws/index.js:10:21","invoke (/var/runtime/node_modules/awslambda/index.js:288:20)","InvokeManager.start (/var/runtime/node_modules/awslambda/index.js:151:9)","Object.awslambda.waitForInvoke (/var/runtime/node_modules/awslambda/index.js:499:52)"]}
```

Any thoughts?

Thanks!

**Comment by dougmoscrop** [comment]
I'm unfamiliar with this! Any good links to read?

**Comment by adamwilbert** [comment]
https://serverless.com/framework/docs/providers/aws/events/apigateway#using-asyncronous-integration

Is really the only documentaiton.  it's essentially just to allow offline processing so the lambda doesn't immediately stop running when you send a response.

When i run the lambda locally through serverless-offline, path is defined and can be sliced normally, but I'm getting this issue where slice is not a function once i deploy the lambda...

**Comment by dougmoscrop** [comment]
any chance you can capture the stringifed event/context and paste it here (as anonymized as need be)? it's probably just a case of needing to pick someProperty || anotherProperty instead.  Normally I'd just deploy and test it myself but I'm a bit swamped

**Comment by adamwilbert** [comment]
Unfortunately that's a bit difficult because of the pipeline this function lives in...

Not really sure how it's possible, as your code looks like it defaults to path as '/' in case of an issue...

**Comment by dougmoscrop** [comment]
Which version of serverless-http do you have deployed?

**Comment by adamwilbert** [comment]
I was on 1.9.1 but upgraded to 2.0.1 in an attempt to solve the issue

**Comment by adamwilbert** [comment]
Any thoughts on a way to resolve this?  the only workaround i've found is bypassing serverless-http entirely, not ideal for our setup.

**Comment by dougmoscrop** [comment]
You can `npm link` in the module and then debug, which is what I will do at some point but I just don't have time right now

**Comment by adamwilbert** [comment]
I'll fork and make some changes (the problem is i can't recreate locally, so linking does do a whole lot of good at the moment) and then submit a PR when i can get it straightened out.

**Comment by dougmoscrop** [comment]
even if you just link in a copy with some console statements that can help!

**Comment by adamwilbert** [comment]
Path is somewhere along the way getting set to `{}` so your check--`event.path || '/'` is bringing in the empty object and attempting to split it.

**Comment by swbullis** [comment]
For those running into this issue, I did manage to get a workaround for my case.  I was getting the same error on 3 of my Functions but the others were fine.  

I did two things 
* Upgraded 2.0.1 (Which may not have been necessary)
* Removed `integration: lambda` from my function

`
function:
  sessionCancel:
    handler: server/src/handlers/session.cancel
    events:
      - http:
        path: api/session/cancel
        method: post
        integration: lambda
        cors: true
`

Hopefully this helps someone out.

**Comment by EdisonHarada** [comment]
I need "integration: lambda" to be able to send custom headers in the response. So you only remove it if you don't need it.

@dougmoscrop 
Here's the event that I'm receiving (I'm using serverless-offline to simulate it locally):
```
{
  "body": {},
  "method": "GET",
  "principalId": "offlineContext_authorizer_principalId",
  "headers": {
    "Host": "localhost:3000",
    "Connection": "keep-alive",
    "Pragma": "no-cache",
    "Cache-Control": "no-cache",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36",
    "Accept": "image/webp,image/apng,image/*,*/*;q=0.8",
    "Referer": "http://localhost:3000/",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-GB,en;q=0.9,pt-BR;q=0.8,pt;q=0.7,en-US;q=0.6",
    "Cookie": "_ga=GA1.1.1105800980.1556334775; _gid=GA1.1.208687299.1564240328"
  },
  "query": {},
  "path": {
    "proxy": "favicon.ico"
  },
  "identity": {
    "accountId": "offlineContext_accountId",
    "apiKey": "offlineContext_apiKey",
    "caller": "offlineContext_caller",
    "cognitoAuthenticationProvider": "offlineContext_cognitoAuthenticationProvider",
    "cognitoAuthenticationType": "offlineContext_cognitoAuthenticationType",
    "sourceIp": "127.0.0.1",
    "user": "offlineContext_user",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36",
    "userArn": "offlineContext_userArn"
  },
  "stageVariables": {},
  "isOffline": true
}
```

If I remove the "integration: lambda" I have this event:
```
{
  "body": null,
  "headers": {
    "Host": "localhost:3000",
    "Connection": "keep-alive",
    "Cache-Control": "max-age=0",
    "Upgrade-Insecure-Requests": "1",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-GB,en;q=0.9,pt-BR;q=0.8,pt;q=0.7,en-US;q=0.6",
    "Cookie": "_ga=GA1.1.1105800980.1556334775; _gid=GA1.1.208687299.1564240328"
  },
  "httpMethod": "GET",
  "multiValueHeaders": {
    "Host": [
      "localhost:3000"
    ],
    "Connection": [
      "keep-alive"
    ],
    "Cache-Control": [
      "max-age=0"
    ],
    "Upgrade-Insecure-Requests": [
      "1"
    ],
    "User-Agent": [
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36"
    ],
    "Accept": [
      "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3"
    ],
    "Accept-Encoding": [
      "gzip, deflate, br"
    ],
    "Accept-Language": [
      "en-GB,en;q=0.9,pt-BR;q=0.8,pt;q=0.7,en-US;q=0.6"
    ],
    "Cookie": [
      "_ga=GA1.1.1105800980.1556334775; _gid=GA1.1.208687299.1564240328"
    ]
  },
  "multiValueQueryStringParameters": null,
  "path": "/",
  "pathParameters": null,
  "queryStringParameters": null,
  "requestContext": {
    "accountId": "offlineContext_accountId",
    "apiId": "offlineContext_apiId",
    "authorizer": {
      "principalId": "offlineContext_authorizer_principalId"
    },
    "httpMethod": "GET",
    "identity": {
      "accountId": "offlineContext_accountId",
      "apiKey": "offlineContext_apiKey",
      "caller": "offlineContext_caller",
      "cognitoAuthenticationProvider": "offlineContext_cognitoAuthenticationProvider",
      "cognitoAuthenticationType": "offlineContext_cognitoAuthenticationType",
      "cognitoIdentityId": "offlineContext_cognitoIdentityId",
      "cognitoIdentityPoolId": "offlineContext_cognitoIdentityPoolId",
      "sourceIp": "127.0.0.1",
      "user": "offlineContext_user",
      "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36",
      "userArn": "offlineContext_userArn"
    },
    "protocol": "HTTP/1.1",
    "requestId": "offlineContext_requestId_cjylp0a440001o8vvhn3mbjj5",
    "requestTimeEpoch": 1564241510423,
    "resourceId": "offlineContext_resourceId",
    "resourcePath": "/",
    "stage": "development"
  },
  "resource": "/",
  "stageVariables": null,
  "isOffline": true
}
```

**Comment by dougmoscrop** [comment]
Maybe fixed by https://github.com/dougmoscrop/serverless-http/pull/111

**Comment by dougmoscrop** [comment]
Can you try off master (just merged that PR) and let me know if its fixed?

**Comment by larswww** [comment]
Hi Doug,

Thank you for the support on this issue! Am having the same problem. Tried updating to 2.0.2 and a few different serverless configs. 

Basically, am trying to have a POST endpoint respond immediately, then want the lambda to continue running doing a bunch of async stuff. 

I get this same error when setting "async: true" in serverless.yml
If i remove that, and use context.callBackWaitsForEmptyEventLopp = false, I get no error, but the lambda doesn't continue running past my async stuff. 

Works locally either way using serverless offline.  Can share source by adding you to a private repo if you'd like?

**Comment by dougmoscrop** [comment]
Sure, also can you try from master? I am yet to release the latest but will soon, but for now just change your package.json version to `github:dougmoscrop/serverless-http#5813aff1f7764` and see if it still happens

**Comment by dougmoscrop** [comment]
Should be fixed in 2.1.0 including support (ish -- try/lemme know) for async/sync lambda (non-proxy) integration

**Comment by larswww** [comment]
Thanks Doug! Using 2.3.0 and the errors are gone, but the lambda ends once it encounters the first await in the async function. 

why do you say "non-proxy" integration?
